wandb/framework/tensorboard/monkeypatch.py

```diff
@@ -0,0 +1,64 @@
+"""
+monkeypatch: patch code to add tensorboard hooks
+"""
+
+import wandb
+
+
+TENSORBOARD_C_MODULE = "tensorflow.python.ops.gen_summary_ops"
+TENSORBOARD_PYTORCH_MODULE = "tensorboard.summary.writer.event_file_writer"
+
+
+def patch(save=None, tensorboardX=None, pytorch=None):
+    if len(wandb.patched["tensorboard"]) > 0:
+        raise ValueError(
+            "Tensorboard already patched, remove sync_tensorboard=True from wandb.init or only call wandb.tensorboard.patch once."
+        )
+
+    c_writer = wandb.util.get_module(TENSORBOARD_C_MODULE)
+    tb_writer = wandb.util.get_module(TENSORBOARD_PYTORCH_MODULE)
+
+    if c_writer:
+        _patch_tensorflow2(writer=c_writer, module=TENSORBOARD_C_MODULE, save=save)
+    elif tb_writer:
+        _patch_nontensorflow(
+            writer=tb_writer, module=TENSORBOARD_PYTORCH_MODULE, save=save
+        )
+    else:
+        raise ValueError("Unsupported tensorboard configuration")
+
+
+def _patch_tensorflow2(writer, module, save=None):
+    # This configures TensorFlow 2 style Tensorboard logging
+    old_csfw_func = writer.create_summary_file_writer
+
+    def new_csfw_func(*args, **kwargs):
+        logdir = (
+            kwargs["logdir"].numpy().decode("utf8")
+            if hasattr(kwargs["logdir"], "numpy")
+            else kwargs["logdir"]
+        )
+        _notify_tensorboard_logdir(logdir, save=save)
+        return old_csfw_func(*args, **kwargs)
+
+    writer.orig_create_summary_file_writer = old_csfw_func
+    writer.create_summary_file_writer = new_csfw_func
+    wandb.patched["tensorboard"].append([module, "create_summary_file_writer"])
+
+
+def _patch_nontensorflow(writer, module, save=None):
+    # This configures non-TensorFlow Tensorboard logging
```

Comment:
non-TensorFlow or TensorFlow < 1.4x I believe.

Reply:
Or better, PyTorch, TensorboardX, or TensorFlow < 1.4x logging.

Reply:
I take both of these back as we're not actually patching old tensorflow with this, which I think is fine for initial release.

Reply:
yep.  i think we can make old tensorflow work with the watcher method, but there are more tricks needed like running event compat conversions.  Would like to defer a bit on this